sparsify/trainer.py

```diff
@@ -702,6 +712,50 @@ def save_best(self, avg_loss: float | dict[str, float]):
         if dist.is_initialized():
             dist.barrier()
 
+    @torch.no_grad()
+    def evaluate(self, eval_dataset, step: int):
+        """Run eval set through spliced vs unspliced model,
+        compute mean KL divergence."""
+        if eval_dataset is None:
+            return
+        # Only rank 0 should run evaluation
+        rank_zero = not dist.is_initialized() or dist.get_rank() == 0
+        if not rank_zero:
+            return
+        # only rank 0 does eval to avoid duplication
+
+        device = self.model.device
+        dl = DataLoader(eval_dataset, batch_size=self.cfg.batch_size, shuffle=False)
+
+        self.model.eval()
+
+        total_kl, total_count = 0.0, 0
+
+        # === Run inference twice per batch ===
+        for batch in tqdm(dl, desc=f"Eval@{step}", disable=False):
+            x = batch["input_ids"].to(device)
+            # 1. Unspliced logits
+            logits_unspliced = self.model(x).logits
+
+            # 2. Spliced logits (with SAE edits)
+            with edit_for_generation(
+                self.model, self.cfg.hookpoints, self.saes, device=device
+            ):
+                logits_spliced = self.model(x).logits
+
+            # === Compute KL divergence ===
+            kl = kl_divergence(logits_spliced, logits_unspliced, dim=-1)
+            total_kl += kl.sum().item()
+            total_count += kl.numel()
+        mean_kl = total_kl / total_count if total_count > 0 else float("nan")
+
+        # Only rank 0 reaches here, so no further rank checks required
```

Comment:
This is also an unnecessary comment. Agents generally add a lot of unnecessary comments to justify their design decisions to the user in the moment but they don't usually add value to the final code